XiEditor/XiDocumentController.swift

```diff
@@ -0,0 +1,51 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+import Cocoa
+
+class XiDocumentController: NSDocumentController {
+
+    fileprivate var lock = UnfairLock()
+    /// Lookup used when routing RPCs to views.
+    fileprivate var openViews = [ViewIdentifier: Document]()
+    
+    /// Returns the document corresponding to this `ViewIdentifier`, if it exists.
+    func documentForViewIdentifier(_ viewIdentifier: ViewIdentifier) -> Document? {
+        lock.lock()
+        defer { lock.unlock() }
+        return openViews[viewIdentifier]
+    }
+    
+    /// Associates the given view with the given identifier. Should only be called when
+    /// the `ViewIdentifier` is first set.
+    func setIdentifier(_ viewIdentifier: ViewIdentifier, forDocument document: Document) {
+        lock.lock()
```

Comment:
Any special reason you're doing explicit lock/unlock here but used defer above?

Reply:
I use defer above because it lets me return without using a temporary value.